x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java

```diff
@@ -216,6 +288,92 @@ public void deleteTrainedModel(String modelId, ActionListener<Boolean> listener)
+        if (pageParams != null) {
+            sourceBuilder.from(pageParams.getFrom()).size(pageParams.getSize());
+        }
+        sourceBuilder.trackTotalHits(true)
+            // we only care about the item id's
+            .fetchSource(TrainedModelConfig.MODEL_ID.getPreferredName(), null);
+
+        IndicesOptions indicesOptions = SearchRequest.DEFAULT_INDICES_OPTIONS;
+        SearchRequest searchRequest = new SearchRequest(InferenceIndexConstants.INDEX_PATTERN)
+            .indicesOptions(IndicesOptions.fromOptions(true,
+                indicesOptions.allowNoIndices(),
+                indicesOptions.expandWildcardsOpen(),
+                indicesOptions.expandWildcardsClosed(),
+                indicesOptions))
+            .source(sourceBuilder);
+
+        executeAsyncWithOrigin(client.threadPool().getThreadContext(),
+            ML_ORIGIN,
+            searchRequest,
+            ActionListener.<SearchResponse>wrap(
+                response -> {
+                    Set<String> foundResourceIds = new LinkedHashSet<>();
+                    long totalHitCount = response.getHits().getTotalHits().value;
+                    for (SearchHit hit : response.getHits().getHits()) {
+                        Map<String, Object> docSource = hit.getSourceAsMap();
+                        if (docSource == null) {
+                            continue;
+                        }
+                        Object idValue = docSource.get(TrainedModelConfig.MODEL_ID.getPreferredName());
+                        if (idValue instanceof String) {
+                            foundResourceIds.add(idValue.toString());
+                        }
+                    }
+                    ExpandedIdsMatcher requiredMatches = new ExpandedIdsMatcher(tokens, allowNoResources);
+                    requiredMatches.filterMatchedIds(foundResourceIds);
+                    if (requiredMatches.hasUnmatchedIds()) {
+                        idsListener.onFailure(ExceptionsHelper.missingTrainedModel(requiredMatches.unmatchedIdsString()));
+                    } else {
+                        idsListener.onResponse(Tuple.tuple(totalHitCount, foundResourceIds));
+                    }
+                },
+                idsListener::onFailure
+            ),
+            client::search);
+
+    }
+
+    private QueryBuilder buildQueryIdExpressionQuery(String[] tokens, String resourceIdField) {
+        BoolQueryBuilder boolQuery = QueryBuilders.boolQuery()
+            .filter(QueryBuilders.termQuery(InferenceIndexConstants.DOC_TYPE.getPreferredName(), TrainedModelConfig.NAME));
+
+        if (Strings.isAllOrWildcard(tokens)) {
+            return boolQuery;
+        }
+        // If the resourceId is not _all or *, we should see if it is a comma delimited string with wild-cards
+        // e.g. id1,id2*,id3
+        BoolQueryBuilder shouldQueries = new BoolQueryBuilder();
```

Comment:
We are expanding the ids in other parts of our codebase. Shouldn't this logic (splitting by comma, wildcards) be reused?

Reply:
@przemekwitek I agree we should have some id expansion class, but the individual pieces are slightly different enough that finding the appropriate abstraction would take some refactoring. I would rather not do that type of refactoring inside of a feature branch..